syntax = "proto3";
package dts.proto;;

import "google/protobuf/struct.proto";

enum TaskState {
  PENDING = 0;
  RUNNING = 1;
  SUCCESS = 2;
  FAILED = 3;
  TIMEOUT = 4;
  CANCELLED = 5;
}

message Resource {
  double cpu_core = 1;
  uint64 mem_mb = 2;
}

message Shard {
  uint32 shard_id = 1;
  uint32 total_shards = 2;
}

message Task {
  string task_id = 1;
  string client_id = 2;
  uint32 priority = 3;
  TaskState state = 4;
  string func_name = 5;
  google.protobuf.Struct func_params = 6;
  Resource required = 7;
  Shard shard = 8;
  uint32 timeout_ms = 9;
  uint32 max_retry = 10;
  uint32 retry_count = 11;
  int64 submit_ts = 12;
  int64 start_ts = 13;
  int64 finish_ts = 14;
  google.protobuf.Struct result = 15;
  string error_msg = 16;
}

message TaskResponse {
  Task task = 1;
}

message CancelRequest {
  string task_id = 1;
}

message CancelResponse {
  bool success = 1;
}

message QueryRequest {
  string task_id = 1;
}

message SubscribeRequest {
  string client_id = 1;
}

message TaskResult {
  Task task = 1;
}

service TaskService {
  rpc SubmitTask(Task) returns (TaskResponse) {}
  rpc CancelTask(CancelRequest) returns (CancelResponse) {}
  rpc QueryStatus(QueryRequest) returns (Task) {}
  rpc ListenResults(SubscribeRequest) returns (stream TaskResult) {}
}